F# plaintext printing documentation (#19632)
dotnet/docs: docs/fsharp/language-reference/fsharp-interactive-options.md

---

> title: Interactive Options

Reviewer: ```suggestion
title: Interactive options
```

description: Learn about the command-line options supported by F# Interactive, fsi.exe.

> ms.date: 22/07/2030

Reviewer: ```suggestion
ms.date: 22/07/2020
```

---

> # F# Interactive Options

Reviewer: ```suggestion
# F# Interactive options
```

This topic describes the command-line options supported by F# Interactive, `fsi.exe`. F# Interactive accepts many of the same command line options as the F# compiler, but also accepts some additional options.

> ## Using F# Interactive for Scripting

Reviewer: ```suggestion
## Using F# Interactive for scripting
```

F# Interactive, `dotnet fsi`, can be launched interactively, or it can be launched from the command line to run a script. The command line syntax is

> ```console

Reviewer: ```suggestion
```dotnetcli
```

> > dotnet fsi [options] [ script-file [arguments] ]

Reviewer: ```suggestion
dotnet fsi [options] [ script-file [arguments] ]
```

|**--warnaserror**[**+**&#124;**-**]|Same as the **fsc.exe** compiler option. For more information, see [Compiler Options](compiler-options.md).|
|**--warnaserror**[**+**&#124;**-**]:**&lt;int-list&gt;**|Same as the **fsc.exe** compiler option. For more information, see [Compiler Options](compiler-options.md).|

## F# Interactive Structured Printing

F# Interactive (`dotnet fsi`) uses an extended version of [structured plain text formatting](plaintext-formatting.md) to
report values.

> 1. All features of `%A` plaintext formatting are supported, and some are additionally customizable (see below).

Reviewer 2: ```suggestion
1. All features of `%A` plain text formatting are supported, and some are additionally customizable.
```
We tend not to use above/below/left/right terminology in docs

> 2. Printing is colorized if colours are aupported by the output console.

Reviewer 2: ```suggestion
2. Printing is colorized if colors are supported by the output console.
```

3. A limit is placed on the length of strings shown, unless you explicitly evaluate that string.

4. A set of user-definable settings are available via the `fsi` object.

The available settings to customize plain text printing for reported values are:

```fsharp
open System.Globalization

> fsi.FormatProvider <- CultureInfo("de-DE")  // control the default culture for primitives
> 
> fsi.PrintWidth <- 120        // control the width used for structured printing
> 
> fsi.PrintDepth <- 10         // control the maximum depth of nested printing
> 
> fsi.PrintLength <- 10        // control the length of lists and arrays
> 
> fsi.PrintSize <- 100         // control the maximum overall object count
> 
> fsi.ShowProperties <- false  // control whether properties of .NET objects are shown by default
> 
> fsi.ShowIEnumerable <- false // control whether sequence values are expanded by default
> 
> fsi.ShowDeclarationValues <- false // control whether values are shown for declaration outputs

Reviewer 2: ```suggestion
fsi.FormatProvider <- CultureInfo("de-DE")  // control the default culture for primitives

fsi.PrintWidth <- 120        // Control the width used for structured printing

fsi.PrintDepth <- 10         // Control the maximum depth of nested printing

fsi.PrintLength <- 10        // Control the length of lists and arrays

fsi.PrintSize <- 100         // Control the maximum overall object count

fsi.ShowProperties <- false  // Control whether properties of .NET objects are shown by default

fsi.ShowIEnumerable <- false // Control whether sequence values are expanded by default

fsi.ShowDeclarationValues <- false // Control whether values are shown for declaration outputs
```

```

### Customizing F# Interactive structured printing with `AddPrinter` and `AddPrintTransformer`

Printing in F# Interactive outputs can be customized by using `fsi.AddPrinter` and `fsi.AddPrintTransformer`.
The first gives text to replace the printing of an object, the second returns a surrogate object to display

> instead. For example, in F# Interactive

Reviewer 2: ```suggestion
instead. For example, consider the following F# code:
```

```fsharp
open System

fsi.AddPrinter<DateTime>(fun dt -> dt.ToString("s"))

type DateAndLabel =
  { Date : DateTime
    Label  : string  }

let newYearsDay1999 =
  { Date = DateTime(1999, 1, 1)
    Label = "New Year" }
```

> gives output that uses the specific date time format:

Reviewer 2: ```suggestion
If you execute the example in F# Interactive, it outputs based on the formatting option set. In this case, it affects the formatting of date and time:
```

```console
type DateAndLabel =
  { Date: DateTime
    Label: string }
val newYearsDay1999 : DateAndLabel = { Date = 1999-01-01T00:00:00
                                       Label = "New Year" }
```

`fsi.AddPrintTransformer` can be used to give a surrogate object for printing:

```fsharp
type MyList(values: int list) =
    member x.Values = values

fsi.AddPrintTransformer(fun (x:MyList) -> box x.Values)

let x = MyList([1..10])
```

> gives

Reviewer 2: ```suggestion
This outputs:
```

```console
val x : MyList = [1; 2; 3; 4; 5; 6; 7; 8; 9; 10]
```

> If the transformer function passed to `fsi,AddPrintTransformer` returns `null` then the print transformer is ignored.

Reviewer 2: ```suggestion
If the transformer function passed to `fsi.AddPrintTransformer` returns `null` then the print transformer is ignored.
```

> This can be used to filter any input value by starting with type `obj`.  For example,

Reviewer 2: ```suggestion
This can be used to filter any input value by starting with type `obj`.  For example:
```

```fsharp
fsi.AddPrintTransformer(fun (x:obj) ->
    match x with
    | :? string as s when s = "beep" -> box ["quack"; "quack"; "quack"]
    | _ -> null)

let y = "beep"
```

> gives

Reviewer 2: ```suggestion
This outputs:
```

```console
val y : string = ["quack"; "quack"; "quack"]
```

> ## Related Topics

Reviewer: I won't add too much suggestions on the same thing, but the idea is to use sentence-style casing, I think that's what the current style-guide says.

```suggestion
## Related topics
```

```fsharp
open System.Globalization

fsi.FormatProvider <- CultureInfo("de-DE")  // control the default culture for primitives

fsi.PrintWidth <- 120        // Control the width used for structured printing

fsi.PrintDepth <- 10         // Control the maximum depth of nested printing

fsi.PrintLength <- 10        // Control the length of lists and arrays

fsi.PrintSize <- 100         // Control the maximum overall object count

fsi.ShowProperties <- false  // Control whether properties of .NET objects are shown by default

fsi.ShowIEnumerable <- false // Control whether sequence values are expanded by default

fsi.ShowDeclarationValues <- false // Control whether values are shown for declaration outputs
```

### Customize with `AddPrinter` and `AddPrintTransformer`

Printing in F# Interactive outputs can be customized by using `fsi.AddPrinter` and `fsi.AddPrintTransformer`.
The first function gives text to replace the printing of an object. The second function returns a surrogate object to display
instead. For example, consider the following F# code:

```fsharp
open System

fsi.AddPrinter<DateTime>(fun dt -> dt.ToString("s"))

> type DateAndLabel =
>   { Date : DateTime
>     Label  : string  }
> 
> let newYearsDay1999 =
>   { Date = DateTime(1999, 1, 1)
>     Label = "New Year" }

Reviewer 2: ```suggestion
type DateAndLabel =
    { Date: DateTime
      Label: string  }

let newYearsDay1999 =
    { Date = DateTime(1999, 1, 1)
      Label = "New Year" }
```

```

If you execute the example in F# Interactive, it outputs based on the formatting option set. In this case, it affects the formatting of date and time:

```console
type DateAndLabel =
  { Date: DateTime
    Label: string }
val newYearsDay1999 : DateAndLabel = { Date = 1999-01-01T00:00:00
                                       Label = "New Year" }
```

`fsi.AddPrintTransformer` can be used to give a surrogate object for printing:

```fsharp
type MyList(values: int list) =

>     member x.Values = values

Reviewer 2: ```suggestion
    member _.Values = values
```
